### PR TITLE
Added Labels as a Selection Criterion

### PR DESCRIPTION
**Pull request by nishaq503**
Selects clusters that have an overwhelming majority of points from one class.

feat: label based selection criteria

### CODE REVIEW

File: pyclam/criterion.py [hunk 1]
```diff
@@ -337,3 +337,36 @@ def _select(root: Cluster, upper: float, lower: float, grace_depth: int) -> Set[
...
+
+
+class Labels(SelectionCriterion):
+    def __init__(self, labels: Dict[int, int]):
+        """ Uses a given dictionary of index -> label to select optimal clusters.
+
+        During a BFT of the tree, when a cluster has an overwhelming majority of points with the same label,
+        that cluster is selected.
+        If some data points are unlabelled, then when a cluster contains no labelled points, it too is selected.
+        """
```
Comment (thoward27) [inline pyclam/criterion.py:349]: We don't usually document in `__init__`, usually docstrings go directly on the class object

File: pyclam/criterion.py [hunk 2]
```diff
@@ -337,3 +337,36 @@ def _select(root: Cluster, upper: float, lower: float, grace_depth: int) -> Set[
             inactive = {child for cluster in inactive for child in cluster.children}
 
         return selected
+
+
+class Labels(SelectionCriterion):
+    def __init__(self, labels: Dict[int, int]):
+        """ Uses a given dictionary of index -> label to select optimal clusters.
+
+        During a BFT of the tree, when a cluster has an overwhelming majority of points with the same label,
```
Comment (thoward27) [inline pyclam/criterion.py:346]: What is an overwhelming majority? Can we parameterize that?

File: pyclam/criterion.py [hunk 3]
```diff
@@ -337,3 +337,36 @@ def _select(root: Cluster, upper: float, lower: float, grace_depth: int) -> Set[
...
+            else:
+                labels: List[int] = [self.labels[p] for p in cluster.argpoints if p in self.labels]
+                if len(labels) > 0:
+                    max_fraction = max(dict(Counter(labels)).values()) / cluster.cardinality
+                    if max_fraction > self.threshold:
+                        selected.add(cluster)
+                    else:
+                        frontier.update(cluster.children)
+                else:
+                    # TODO: Is this really the best idea?
```
Comment (thoward27) [inline pyclam/criterion.py:370]: Seems like the only good path. We can't expect that future children will suddenly have more labels. Unless you just take these types of situations down to singletons? What other ideas did you have?

**Review by thoward27 — CHANGES_REQUESTED** [review]
I'd like to see the overwhelming majority bit better defined in our documentation and potentially parameterized if possible.

File: pyclam/criterion.py (hunk 1 above)
Comment (nishaq503) [reply pyclam/criterion.py]: moved

File: pyclam/criterion.py (hunk 2 above)
Comment (nishaq503) [reply pyclam/criterion.py]: It was already parameterized in self.threshold. I added an explanation to the doc-string.

File: pyclam/criterion.py (hunk 3 above)
Comment (nishaq503) [reply pyclam/criterion.py]: reverting to other selection criteria?

File: pyclam/criterion.py [hunk 4]
```diff
@@ -392,3 +393,40 @@ def __call__(self, root: Cluster) -> Set[Cluster]:
...
+class Labels(SelectionCriterion):
+    """ Uses a given dictionary of index -> label to select optimal clusters.
+
+    During a BFT of the tree, when a cluster has an overwhelming majority of points with the same label,
+    that cluster is selected.
+    If some data points are unlabelled, then when a cluster contains no labelled points, it too is selected.
+
+    The 'overwhelming majority' fraction is parameterized by the fraction of the minority class.
+    Let f be the fraction of the smallest minority class in all the data.
+    Then the threshold for determining an overwhelming majority is (1 - f / 10).
```
Comment (thoward27) [inline pyclam/criterion.py:407]: I meant you should parameterize the magical `10`, unless there's a good reason for it to be 10?

**Review by thoward27 — COMMENTED** [review]
I don't think our definitions of parameterized quite lined up, let me know your thoughts

File: pyclam/criterion.py (hunk 4 above)
Comment (nishaq503) [reply pyclam/criterion.py]: 10 is a very good number here so I am going to choose an appropriate default value for the argument

**Review by thoward27 — APPROVED** [review]
LGTM 🎉

**Pull request merged**